File: typescript/resistor-color-trio/resistor-color-trio.ts
const Colors = [
  "black",
  "brown",
  "red",
  "orange",
  "yellow",
  "green",
  "blue",
  "violet",
  "grey",
  "white",
];

const PREFIXES = [
  { base: 1_000_000_000, prefix: "giga" },
  { base: 1_000_000, prefix: "mega" },
  { base: 1_000, prefix: "kilo" },
];

export function decodedResistorValue(colors: string[]): string {
  const decodedOhms =
    (Colors.indexOf(colors[0]) * 10 + Colors.indexOf(colors[1])) *
    Math.pow(10, Colors.indexOf(colors[2]));

  const encoding = PREFIXES.find((ohm) => ohm.base <= decodedOhms);
  const base = encoding?.base || 1;
  const prefix = encoding?.prefix || "";
  return `${decodedOhms / base} ${prefix}ohms`;
}

File: prolog/resistor-color-trio/resistor_color_trio.pl
:- use_module(library(clpfd)).

unit_quantity(Value, Quantity, "gigaohms") :-
    Value #> 1000000000, Quantity #= Value // 1000000000, !.

unit_quantity(Value, Quantity, "megaohms") :-
    Value #> 1000000, Quantity #= Value // 1000000, !.

unit_quantity(Value, Quantity, "kiloohms") :-
    Value #> 1000, Quantity #= Value // 1000, !.

unit_quantity(Value, Value, "ohms").

label(Colors, Label) :-
    value(Colors, Value),
    unit_quantity(Value, Quantity, Unit),
    format(string(Label), "~d ~w", [Quantity, Unit]).

value([Color1, Color2, Color3|_], Value) :-
    color_code(Color1, Tens),
    color_code(Color2, Digit),
    color_code(Color3, Zeros),
    Value #= (Tens * 10 + Digit) * (10 ^ Zeros).

color_code(Color, Code) :-
    colors(Colors),
    nth0(Code, Colors, Color), !.

colors(Colors) :-
    Colors = ["black", "brown", "red", "orange", "yellow", "green", "blue", "violet", "grey", "white"].
